FirebaseCombineSwift/Sources/Storage/StorageReference.swift

```diff
@@ -0,0 +1,306 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#if canImport(Combine) && swift(>=5.0) && canImport(FirebaseStorage)
+
+  import Combine
+  import FirebaseStorage
+
+  @available(swift 5.0)
+  @available(macOS 10.15, iOS 13, watchOS 6, tvOS 13, *)
+  extension StorageReference {
+    // MARK: - Uploads
+
+    /// Asynchronously uploads data to the currently specified FIRStorageReference.
+    /// This is not recommended for large files, and one should instead upload a file from disk.
+    ///
+    /// The publisher will emit events on the **main** thread.
+    ///
+    /// - Parameters:
+    ///   - uploadData: The Data to upload.
+    ///   - metadata: metadata `StorageMetadata` containing additional information (MIME type, etc.)
+    ///
+    /// - Returns: A publisher emitting a `StorageMetadata` instance. The publisher will emit on the *main* thread.
+    @discardableResult
+    public func putDataPublisher(_ data: Data,
+                                 metadata: StorageMetadata?)
+      -> AnyPublisher<StorageMetadata, Error> {
+      let subject = PassthroughSubject<StorageMetadata, Error>()
```

Comment:
Shouldn't this be a Future instead? We call once, and receive either an error or a metadata object.

Reply:
@peterfriese I've used `PassthroughSubject` to manage the returned task. I couldn't find any other alternatives to wire the cancellation of a `Future` asynchronous code.

Reply:
I've sent a PR (https://github.com/ziadtamim/firebase-ios-sdk/pull/1) to your branch with a suggestion how to us Futures for this. The test passes.

Reply:
Thanks @peterfriese